### 필요라이브러리 설치

In [1]:
%pip install --upgrade nbformat
%pip install plotly

  Using cached jsonschema-4.22.0-py3-none-any.whl.metadata (8.2 kB)
  Using cached jsonschema_specifications-2023.12.1-py3-none-any.whl.metadata (3.0 kB)
  Using cached referencing-0.35.1-py3-none-any.whl.metadata (2.8 kB)
  Using cached rpds_py-0.18.1-cp312-none-win_amd64.whl.metadata (4.2 kB)
   ---------------------------------------- 0.0/78.5 kB ? eta -:--:--
   ----- ---------------------------------- 10.2/78.5 kB ? eta -:--:--
   ----- ---------------------------------- 10.2/78.5 kB ? eta -:--:--
   -------------------- ------------------- 41.0/78.5 kB 393.8 kB/s eta 0:00:01
   ---------------------------------------- 78.5/78.5 kB 487.2 kB/s eta 0:00:00
Using cached jsonschema-4.22.0-py3-none-any.whl (88 kB)
Using cached jsonschema_specifications-2023.12.1-py3-none-any.whl (18 kB)
Using cached referencing-0.35.1-py3-none-any.whl (26 kB)
Using cached rpds_py-0.18.1-cp312-none-win_amd64.whl (209 kB)
Note: you may need to restart the kernel to use updated packages.
Note: you may nee

### Importing Libraries

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px

In [3]:
import warnings
warnings.filterwarnings("ignore")

### Data Loading

In [4]:
df = pd.read_csv(r"C:\Users\USER\Desktop\실습\개인작업\cybersecurity_attacks.csv")

### EDA (Exploratory Data Analysis)

In [5]:
df.head().T

0  \
Timestamp                                             2023-05-30 06:33:58   
Source IP Address                                           103.216.15.12   
Destination IP Address                                       84.9.164.252   
Source Port                                                         31225   
Destination Port                                                    17616   
Protocol                                                             ICMP   
Packet Length                                                         503   
Packet Type                                                          Data   
Traffic Type                                                         HTTP   
Payload Data            Qui natus odio asperiores nam. Optio nobis ius...   
Malware Indicators                                           IoC Detected   
Anomaly Scores                                                      28.67   
Alerts/Warnings                                                       NaN   
Attack Type                                                       Malware   
Attack Signature                                          Known Pattern B   
Action Taken                                                       Logged   
Severity Level                                                        Low   
User Information                                            Reyansh Dugal   
Device Information      Mozilla/5.0 (compatible; MSIE 8.0; Windows NT ...   
Network Segment                                                 Segment A   
Geo-location Data                                      Jamshedpur, Sikkim   
Proxy Information                                            150.9.97.135   
Firewall Logs                                                    Log Data   
IDS/IPS Alerts                                                        NaN   
Log Source                                                         Server   

                                                                        1  \
Timestamp                                             2020-08-26 07:08:30   
Source IP Address                                          78.199.217.198   
Destination IP Address                                     66.191.137.154   
Source Port                                                         17245   
Destination Port                                                    48166   
Protocol                                                             ICMP   
Packet Length                                                        1174   
Packet Type                                                          Data   
Traffic Type                                                         HTTP   
Payload Data            Aperiam quos modi officiis veritatis rem. Omni...   
Malware Indicators                                           IoC Detected   
Anomaly Scores                                                       51.5   
Alerts/Warnings                                                       NaN   
Attack Type                                                       Malware   
Attack Signature                                          Known Pattern A   
Action Taken                                                      Blocked   
Severity Level                                                        Low   
User Information                                               Sumer Rana   
Device Information      Mozilla/5.0 (compatible; MSIE 8.0; Windows NT ...   
Network Segment                                                 Segment B   
Geo-location Data                                      Bilaspur, Nagaland   
Proxy Information                                                     NaN   
Firewall Logs                                                    Log Data   
IDS/IPS Alerts                                                        NaN   
Log Source                                                       Firewall   

                                                                  

In [6]:
# List Columns
df.columns

Index(['Timestamp', 'Source IP Address', 'Destination IP Address',
       'Source Port', 'Destination Port', 'Protocol', 'Packet Length',
       'Packet Type', 'Traffic Type', 'Payload Data', 'Malware Indicators',
       'Anomaly Scores', 'Alerts/Warnings', 'Attack Type', 'Attack Signature',
       'Action Taken', 'Severity Level', 'User Information',
       'Device Information', 'Network Segment', 'Geo-location Data',
       'Proxy Information', 'Firewall Logs', 'IDS/IPS Alerts', 'Log Source'],
      dtype='object')

In [7]:
# Shape of data
print(f"There are {df.shape[0]}, row and {df.shape[1]} columns in the dataset")

There are 40000, row and 25 columns in the dataset


In [8]:
# dataset Info
df.info

<bound method DataFrame.info of                  Timestamp Source IP Address Destination IP Address  \
0      2023-05-30 06:33:58     103.216.15.12           84.9.164.252   
1      2020-08-26 07:08:30    78.199.217.198         66.191.137.154   
2      2022-11-13 08:23:25      63.79.210.48          198.219.82.17   
3      2023-07-02 10:38:46     163.42.196.10        101.228.192.255   
4      2023-07-16 13:11:07     71.166.185.76        189.243.174.238   
...                    ...               ...                    ...   
39995  2023-05-26 14:08:42      26.36.109.26         121.100.75.240   
39996  2023-03-27 00:38:27      17.21.163.81         196.108.134.78   
39997  2022-03-31 01:45:49     162.35.217.57            98.107.0.15   
39998  2023-09-22 18:32:38    208.72.233.205         173.79.112.252   
39999  2023-10-10 11:59:52     14.102.21.108           109.198.45.7   

       Source Port  Destination Port Protocol  Packet Length Packet Type  \
0            31225             17616   

### Checking the Missing/Null Values

In [9]:
df.isnull().sum().sort_values(ascending=False)

Alerts/Warnings           20067
IDS/IPS Alerts            20050
Malware Indicators        20000
Firewall Logs             19961
Proxy Information         19851
Attack Type                   0
Geo-location Data             0
Network Segment               0
Device Information            0
User Information              0
Severity Level                0
Action Taken                  0
Attack Signature              0
Timestamp                     0
Source IP Address             0
Anomaly Scores                0
Payload Data                  0
Traffic Type                  0
Packet Type                   0
Packet Length                 0
Protocol                      0
Destination Port              0
Source Port                   0
Destination IP Address        0
Log Source                    0
dtype: int64

- Following Columns have missing values:
- Alerts/Warnings 20067
- IDS/IPS Alerts 20050
- Malware Indicators 20000
- Firewall Logs 19961
- Proxy Information 19851

In [10]:
# Missing Value by Percentage
df.isnull().sum() / len(df) * 100

Timestamp                  0.0000
Source IP Address          0.0000
Destination IP Address     0.0000
Source Port                0.0000
Destination Port           0.0000
Protocol                   0.0000
Packet Length              0.0000
Packet Type                0.0000
Traffic Type               0.0000
Payload Data               0.0000
Malware Indicators        50.0000
Anomaly Scores             0.0000
Alerts/Warnings           50.1675
Attack Type                0.0000
Attack Signature           0.0000
Action Taken               0.0000
Severity Level             0.0000
User Information           0.0000
Device Information         0.0000
Network Segment            0.0000
Geo-location Data          0.0000
Proxy Information         49.6275
Firewall Logs             49.9025
IDS/IPS Alerts            50.1250
Log Source                 0.0000
dtype: float64

### Let's Handle the Missing Values First

In [11]:
# Determine recent activity
df["Alerts/Warnings"] = df["Alerts/Warnings"].apply(lambda x: "yes" if x == "Alert Triggered" else "no")
print(df["Alerts/Warnings"])

0         no
1         no
2        yes
3        yes
4        yes
        ... 
39995     no
39996     no
39997     no
39998    yes
39999    yes
Name: Alerts/Warnings, Length: 40000, dtype: object


- If the Alert Triggered is present, then it's a yes, else it's a no.

In [12]:
df["Malware Indicators"] = df["Malware Indicators"].apply(lambda x: "No Detection" if pd.isna(x) else x)
print(df["Malware Indicators"])

0        IoC Detected
1        IoC Detected
2        IoC Detected
3        No Detection
4        No Detection
             ...     
39995    IoC Detected
39996    IoC Detected
39997    IoC Detected
39998    IoC Detected
39999    No Detection
Name: Malware Indicators, Length: 40000, dtype: object


- If the Malware Indicators is present, then it's a No, else it's a No Detection.

In [13]:
df['Proxy Information'] = df['Proxy Information'].apply(lambda x: 'No proxy' if pd.isna(x) else x)
print(df["Proxy Information"])

0           150.9.97.135
1               No proxy
2         114.133.48.179
3               No proxy
4          149.6.110.119
              ...       
39995           No proxy
39996        60.51.30.46
39997           No proxy
39998       137.76.130.8
39999    112.169.115.139
Name: Proxy Information, Length: 40000, dtype: object


- If Proxy Information is missing, it is assumed that there is no proxy

In [14]:
df['Firewall Logs'] = df['Firewall Logs'].apply(lambda x: 'No Data' if pd.isna(x) else x)
print(df["Firewall Logs"])

0        Log Data
1        Log Data
2        Log Data
3         No Data
4         No Data
           ...   
39995    Log Data
39996    Log Data
39997    Log Data
39998    Log Data
39999    Log Data
Name: Firewall Logs, Length: 40000, dtype: object


- If Firewall Logs is missing, it is assumed that there is no data

In [15]:
df['IDS/IPS Alerts'] = df['IDS/IPS Alerts'].apply(lambda x: 'No Data' if pd.isna(x) else x)
print(df["IDS/IPS Alerts"])

0           No Data
1           No Data
2        Alert Data
3        Alert Data
4        Alert Data
            ...    
39995    Alert Data
39996       No Data
39997    Alert Data
39998       No Data
39999    Alert Data
Name: IDS/IPS Alerts, Length: 40000, dtype: object


- If IDS/IPS Alerts is "No Data", then it means that the alert was not generated by IDS/IPS.

In [16]:
df.isnull().sum().sort_values(ascending=False)

Timestamp                 0
Attack Type               0
IDS/IPS Alerts            0
Firewall Logs             0
Proxy Information         0
Geo-location Data         0
Network Segment           0
Device Information        0
User Information          0
Severity Level            0
Action Taken              0
Attack Signature          0
Alerts/Warnings           0
Source IP Address         0
Anomaly Scores            0
Malware Indicators        0
Payload Data              0
Traffic Type              0
Packet Type               0
Packet Length             0
Protocol                  0
Destination Port          0
Source Port               0
Destination IP Address    0
Log Source                0
dtype: int64

- All Missing Values are removed.

### Let's Explore the Device Information Column

In [17]:
df['Device Information'].value_counts()

Device Information
Mozilla/5.0 (compatible; MSIE 6.0; Windows NT 6.2; Trident/3.0)                                                                                       35
Mozilla/5.0 (compatible; MSIE 5.0; Windows 98; Trident/4.1)                                                                                           34
Mozilla/5.0 (compatible; MSIE 6.0; Windows CE; Trident/4.0)                                                                                           33
Mozilla/5.0 (compatible; MSIE 7.0; Windows NT 6.0; Trident/3.0)                                                                                       31
Mozilla/5.0 (compatible; MSIE 5.0; Windows NT 5.2; Trident/4.1)                                                                                       31
                                                                                                                                                      ..
Mozilla/5.0 (Macintosh; PPC Mac OS X 10_9_2; rv:1.9.2.20) Gecko

In [18]:
# Extract 'Device'
df['Browser'] = df['Device Information'].str.split('/').str[0]
print(df["Browser"])

0        Mozilla
1        Mozilla
2        Mozilla
3        Mozilla
4        Mozilla
          ...   
39995    Mozilla
39996    Mozilla
39997    Mozilla
39998    Mozilla
39999    Mozilla
Name: Browser, Length: 40000, dtype: object


- We've created the Browser column.

In [19]:
import re
# OS and device patterns to search for
patterns = [
    r'Windows',
    r'Linux',
    r'Android',
    r'iPad',
    r'iPod',
    r'iPhone',
    r'Macintosh',
]

def extract_device_or_os(user_agent):
    for pattern in patterns:
        match = re.search(pattern, user_agent, re.I)  # re.I makes the search case-insensitive
        if match:
            return match.group()
    return 'Unknown'  # Return 'Unknown' if no patterns match

# Extract device or OS
df['Device/OS'] = df['Device Information'].apply(extract_device_or_os)
print(df["Device/OS"])

0          Windows
1          Windows
2          Windows
3        Macintosh
4          Windows
           ...    
39995         iPad
39996      Windows
39997      Windows
39998        Linux
39999         iPod
Name: Device/OS, Length: 40000, dtype: object


- We've created the Device/OS column.

In [20]:
df['Browser'].value_counts()

Browser
Mozilla    31951
Opera       8049
Name: count, dtype: int64

In [21]:
df['Device/OS'].value_counts()

Device/OS
Windows      17953
Linux         8840
Macintosh     5813
iPod          2656
Android       1620
iPhone        1567
iPad          1551
Name: count, dtype: int64

- Dropping the Device Information Column

In [22]:
df = df.drop('Device Information', axis = 1)

- All Missing Values are handled, now we can move on to the next step.

In [23]:
def extract_time_features(df, Timestamp):
    # Convert timestamp column to datetime if it's not already
    df[Timestamp] = pd.to_datetime(df[Timestamp])
    
    # Extract time features
    df['Year'] = df[Timestamp].dt.year
    df['Month'] = df[Timestamp].dt.month
    df['Day'] = df[Timestamp].dt.day
    df['Hour'] = df[Timestamp].dt.hour
    df['Minute'] = df[Timestamp].dt.minute
    df['Second'] = df[Timestamp].dt.second
    df['DayOfWeek'] = df[Timestamp].dt.dayofweek
    
    return df

In [24]:
# Assuming df is your DataFrame
# Call the function and store the result in a new DataFrame
new_df = extract_time_features(df, 'Timestamp')

# Check if new columns are created
print(new_df.head())

            Timestamp Source IP Address Destination IP Address  Source Port  \
0 2023-05-30 06:33:58     103.216.15.12           84.9.164.252        31225   
1 2020-08-26 07:08:30    78.199.217.198         66.191.137.154        17245   
2 2022-11-13 08:23:25      63.79.210.48          198.219.82.17        16811   
3 2023-07-02 10:38:46     163.42.196.10        101.228.192.255        20018   
4 2023-07-16 13:11:07     71.166.185.76        189.243.174.238         6131   

   Destination Port Protocol  Packet Length Packet Type Traffic Type  \
0             17616     ICMP            503        Data         HTTP   
1             48166     ICMP           1174        Data         HTTP   
2             53600      UDP            306     Control         HTTP   
3             32534      UDP            385        Data         HTTP   
4             26646      TCP           1462        Data          DNS   

                                        Payload Data  ... Log Source  Browser  \
0  Qui natu

In [ ]:
df.head().T

In [ ]:
df.describe(include = 'object').T

In [ ]:
df.columns

### Let's Explore Each Column

- First We'll Explore the Timestamp Column.

In [ ]:
# Checking the Day Column ploting with plotly
plt = px.histogram(df, x = 'Day', color = 'Malware Indicators', title = 'Number of Malware Attacks by Day')
plt.show()